kythe/docs/rfc/2966.md

```diff
@@ -0,0 +1,331 @@
+
+        6    \n  f    o    o    b    a    r      -- text
+        0x36 0xa 0x66 0x6f 0x6f 0x62 0x61 0x72   -- Unicode (hex)
+
+-   Within this framing, the analyzer communicates with the driver using
+    [JSON-RPC 2.0](http://www.jsonrpc.org/specification) requests and responses,
+    as defined below. The analyzer is the JSON-RPC client, and the driver is the
+    server.
+
+-   Under certain circumstances, the analyzer process **must exit**. In a case
+    where the analyzer process must exit, it must terminate within a grace
+    period. If it does not, the process may be forcibly terminated at any time
+    by the operating environment. The exit code of an analyzer process is
+    ignored. The grace period is defined by the environment, but must be at
+    least 10 seconds.
+
+-   If and after the analyzer closes the output stream, the analyzer process
+    must exit.
+
+-   If and after the driver closes the input stream, the analyzer process must
+    exit.
+
+-   A **corrupt frame** is one in which the length tag does not have the correct
+    form. A frame is corrupt if the newline terminator is missing, the string
+    preceding the newline terminator contains non-digits or is empty, or the
+    stream does ends before the specified number of bytes is delivered.
+
+    If the analyzer receives a corrupt frame from the driver, it must exit. If
+    the driver receives a corrupt frame from the analyzer, it must close the
+    input stream (in which case the analyzer therefore must exit).
+
+### Messages
+
+The following message structures are used by the protocol:
+
+-   `InitRequest`
+
+    ```
+    {
+      "protocol": <protocol>
+    }
+    ```
+
+-   `InitReply`
+
+    ```
+    {
+      "protocol": <protocol>
+    }
+    ```
+
+-   `AnalyzerRequest`
+
+    ```
+    {
+      "language": <language-label>
+    }
+    ```
+
+-   `AnalyzeReply`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "unit": <compilation>
+    }
+    ```
+
+-   `FileRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "path": <file-path>,
+      "digest": <digest>
+    }
+    ```
+
+-   `FileReply`
+
+    ```
+    {
+      "path": <file-path>,
+      "digest": <digest>,
+      "data": <bytes>
+    }
+    ```
+
+-   `OutRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "output": [<bytes>, ...],  -- may be omitted if empty
+      "entries": [<entry>, ...]  -- may be omitted if empty
+    }
+    ```
+
+-   `LogRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "message": <diagnostic>
+    }
+    ```
+
+The basic types referenced above are:
+
+-   `<analysis-id>`: an integer ID assigned by the server (JSON: number).
+-   `<bytes>`: a base64-encoded byte string (JSON: string).
+-   `<compilation>`: a `kythe.proto.CompilationUnit` message (JSON: object).
+-   `<diagnostic>`: a diagnostic message (JSON: string).
+-   `<digest>`: a lower-case hex-coded SHA-256 digest (JSON: string).
```

Comment:
That's not a bad idea, but if we want to do that, we should do it for _all_ the digests in the CompilationUnit and elsewhere. 

This is consistent with what the CU contains now, and to change it would require baking knowledge of the digest into the analyzer implementation, since it is not available from upstream.  I don't think we should make that decision in this doc.